Sources/Atomics/atomics-reference.swift

```diff
@@ -231,3 +233,68 @@ extension AtomicReference
 #endif
 
 }
+
+public struct AtomicTaggedReference<T: AnyObject> {
+    
+    #if swift(>=4.2)
+    @usableFromInline internal var ptr = TaggedOpaqueUnmanagedHelper()
+    #else
+    @_versioned internal var ptr = TaggedOpaqueUnmanagedHelper()
+    #endif
+    
+    public init(_ reference: T? = nil, tag: Int = 0) {
+        initialize(reference, tag: tag)
+    }
+    
+    public mutating func initialize(_ reference: T?, tag: Int) {
+        let u = reference.map(Unmanaged.passRetained)
+        let tp = TaggedOptionalRawPointer(u?.toOpaque(), tag: tag)
+        
+        ptr.initialize(tp)
+    }
+    
+    @discardableResult @inlinable public mutating func swap(_ reference: T?, tag: Int, order: MemoryOrder = .sequential) -> (ref: T?, tag: Int) {
+        let u = reference.map(Unmanaged.passRetained)
+        let tp = TaggedOptionalRawPointer(u?.toOpaque(), tag: tag)
+
+        let pointer = ptr.spinSwap(tp, order)
+        return (pointer.ptr.map(Unmanaged.fromOpaque)?.takeRetainedValue(), pointer.tag)
+    }
+    
+    @inlinable public mutating func take(order: LoadMemoryOrder = .sequential) -> (ref: T?, tag: Int) {
+        let tp = TaggedOptionalRawPointer(nil, tag: 0)
+        let pointer = ptr.spinSwap(tp, MemoryOrder(rawValue: order.rawValue)!)
+        return (tp.ptr.map(Unmanaged<T>.fromOpaque)?.takeRetainedValue(), pointer.tag)
+    }
+    
+    @inlinable public mutating func load(order: LoadMemoryOrder = .sequential) -> (ref: T?, tag: Int) {
+        let tp = ptr.lockAndLoad(order)
+        if let pointer = tp.ptr
+        {
+            assert(ptr.load(.sequential).ptr == UnsafeRawPointer(bitPattern: 0x7) && ptr.load(.sequential).tag == 0)
```

Comment:
This could probably be only one load, compared with a newly-initialized `TaggedOptionalRawPointer`. (Maybe this requires defining == for it, but it still saves one atomic load.)